flang/include/flang/Lower/PFTBuilder.h

```diff
@@ -360,61 +365,155 @@ struct ProgramUnit : ProgramVariant {
+          offset{offset} {}
+    Interval interval{};
+    bool global{};
+    const semantics::Symbol *obj{};
+    std::size_t offset{}; // offset of obj relative to interval
+  };
+
   explicit Variable(const Fortran::semantics::Symbol &sym, bool global = false,
                     int depth = 0)
-      : u{&sym}, depth{depth}, global{global} {}
-  explicit Variable(StoreInterval &&store, bool global = false)
-      : u{std::move(store)}, depth{0}, global{global} {}
+      : var{Nominal(&sym, depth, global)} {}
+  explicit Variable(Interval &&interval, bool global = false)
+      : var{IntervalStore(std::move(interval), global)} {}
+  explicit Variable(IntervalStore &&istore) : var{std::move(istore)} {}
 
+  /// Return the front-end symbol for a nominal variable.
   const Fortran::semantics::Symbol &getSymbol() const {
     assert(hasSymbol());
-    return *std::get<const Fortran::semantics::Symbol *>(u);
+    return *std::get<Nominal>(var).symbol;
   }
 
-  const StoreInterval &getPrimaryStore() const {
-    assert(isPrimaryStore());
-    return std::get<StoreInterval>(u);
+  /// Return the aggregate store.
+  const IntervalStore &getAggregateStore() const {
+    assert(isAggregateStore());
+    return std::get<IntervalStore>(var);
   }
 
-  bool hasSymbol() const {
-    return std::holds_alternative<const Fortran::semantics::Symbol *>(u);
+  /// Return the interval range of an aggregate store.
+  const Interval &getInterval() const {
+    assert(isAggregateStore());
+    return std::get<IntervalStore>(var).interval;
   }
-  bool isPrimaryStore() const { return !hasSymbol(); }
-  bool isGlobal() const { return global; }
-  bool isHeapAlloc() const { return heapAlloc; }
-  bool isPointer() const { return pointer; }
-  bool isTarget() const { return target; }
-  int getDepth() const { return depth; }
 
-  bool isAlias() const { return aliasee; }
-  std::size_t getAlias() const { return aliasOffset; }
+  /// Only nominal variable have front-end symbols.
+  bool hasSymbol() const { return std::holds_alternative<Nominal>(var); }
+
+  /// Is this an aggregate store?
+  bool isAggregateStore() const {
+    return std::holds_alternative<IntervalStore>(var);
+  }
+
+  /// Is this variable a global?
+  bool isGlobal() const {
+    return std::visit([](const auto &x) { return x.global; }, var);
+  }
+
+  bool isHeapAlloc() const {
+    if (auto *s = std::get_if<Nominal>(&var))
+      return s->heapAlloc;
+    return false;
+  }
+  bool isPointer() const {
+    if (auto *s = std::get_if<Nominal>(&var))
+      return s->pointer;
+    return false;
+  }
+  bool isTarget() const {
+    if (auto *s = std::get_if<Nominal>(&var))
+      return s->target;
+    return false;
+  }
+
+  /// An alias(er) is a variable that is part of a EQUIVALENCE that is allocated
+  /// locally on the stack.
+  bool isAlias() const {
+    if (auto *s = std::get_if<Nominal>(&var))
+      return s->aliaser;
+    return false;
+  }
+  std::size_t getAlias() const {
```

Comment:
Maybe getAlias -> getAliasOffset ?

Reply:
That's over specification. It just happens that the alias is identified by an offset. It's not a requirement.